tensorflow/core/common_runtime/bfc_allocator.cc

```diff
@@ -260,6 +262,90 @@ size_t BFCAllocator::RoundedBytes(size_t bytes) {
   return rounded_bytes;
 }
 
+bool BFCAllocator::DeallocateFreeRegions(size_t rounded_bytes) {
+  // Do nothing if garbage collection is off.
+  if (!garbage_collection_) {
+    return false;
+  }
+
+  // Searching for free regions.
+  absl::flat_hash_set<void*> free_region_ptrs;
+  size_t total_free_bytes = 0;
+  for (const AllocationRegion& region : region_manager_.regions()) {
+    ChunkHandle h = region_manager_.get_handle(region.ptr());
```

Comment:
I believe this returns the handle at the beginning of the region. Regions are further divided into bins.

The while loop then iterates over all the chunks in its bin.  If I read this correctly, it does not iterate over all the chunks in the region.  Is this an issue?

Reply:
A `region` contains a linked list of chunks. `Bin` is another data structure recording a set of chunk handles (see `FreeChunkSet`) that have similar sizes to assist search for best fit. So, for each `region`, we can visit all the chunks (owned by the region) through the linked list.

The for loop at line 274 iterates through all the `regions`. For each `region`, we then iterate through all the `chunks` in the `region` following the linked list (i.e., the while loop at line 277).

Checking all the chunks in a region is enough to verify whether the region is free or in use. When we de-allocate the region, we have to remove all its chunks from Bins (line 340.) This makes sure no dangling chunk handles in Bins.